### ISSUE TITLE
Parsing fails with mixture of classes and dataclasses

### ISSUE THREAD

**Issue by williford** [omni-us/jsonargparse#569]
## 🐛 Bug report

Validation fails when there is a union of dataclass and regular classes.

### To reproduce (Updated Example)

Create `cli.py`:

```
from dataclasses import dataclass
from typing import Dict, Optional, Union

from jsonargparse import CLI


@dataclass
class MyData:
    val_a: Optional[str] = None
    val_b: Optional[str] = None


class MyClass:
    def __init__(self, val_1: str, val_2: str):
        self.val_1 = val_1
        self.val_2 = val_2


class MyContainer:
    def __init__(self, value: Union[MyData, MyClass]):
        self.value = value

def command(mycontainer: Optional[MyContainer] =[]) -> None:
    print("Success")


if __name__ == '__main__':
    CLI(command)
```

And config.yaml file:

```
mycontainer:
  class_path: MyContainer
  init_args:
    value:
      class_path: MyClass
      init_args:
        val_1: "A"
        val_2: "B"
```

Run the command:

```
python cli.py --config config.yaml
```

The following error results:

```
usage: cli.py [-h] [--config CONFIG] [--print_config[=flags]] [--mycontainer.help CLASS_PATH_OR_NAME] [--mycontainer MYCONTAINER]
              [--print_shtab {bash,zsh,tcsh}]
error: Validation failed: Parser key "mycontainer":
  Does not validate against any of the Union subtypes
  Subtypes: (<class '__main__.MyContainer'>, <class 'NoneType'>)
  Errors:
    - Problem with given class_path '__main__.MyContainer':
        Parser key "value":
          Does not validate against any of the Union subtypes
          Subtypes: (<class '__main__.MyData'>, <class '__main__.MyClass'>)
          Errors:
            - No action for key "class_path" to set its default.
            - Not a valid subclass of MyClass
              Subclass types expect one of:
              - a class path (str)
              - a dict with class_path entry
              - a dict without class_path but with init_args entry (class path given previously)
          Given value type: <class 'jsonargparse._namespace.Namespace'>
          Given value: Namespace(val_a=None, val_b=None, class_path='MyClass', init_args=Namespace(val_1='A', val_2='B'))
    - Expected a <class 'NoneType'>
  Given value type: <class 'jsonargparse._namespace.Namespace'>
  Given value: Namespace(class_path='__main__.MyContainer', init_args=Namespace(value=Namespace(val_a=None, val_b=None, class_path='MyClass', init_args=Namespace(val_1='A', val_2='B'))))
```

It should print "Success".

If the `Union[MyData, MyClass]` is replaced with `Union[MyClass]`, then it is successful.

### Environment

- jsonargparse version: 4.32.1
- Python version: 3.10.13
- How jsonargparse was installed: ` pip install "jsonargparse[all]>=4.27.7"`
- OS: Ubuntu 24.04 LTS

**Comment by mauvilsa** [comment]
Thank you for reporting!

This might be related to other issues which I am planning to fix some day by removing distinctions between dataclasses and subclasses. But that will take time. Not sure yet if this one makes sense to fix separately.

**Comment by williford** [comment]
Thanks @mauvilsa, Before jsonargparse 4.21.0, dataclasses and subclasses could be treated the same - with giving `class_path` and `init_args` (whether or not that was intended). In these versions it could handle the dataclasses and subclasses being mixed together. Just curious, are you thinking of moving back to allowing that in the future?

**Comment by mauvilsa** [comment]
Yes this should work in the future. Good to know that it worked before. Not sure what changed that makes it fail now. There is no unit test for it, which is why it wasn't noted.

**Comment by mauvilsa** [comment]
This is now implemented in #581. Changed it to enhancement since previously it wasn't officially supported.
